src/solvers/elliptic/linearSolver/PCG_EIGEN.cpp

```diff
@@ -0,0 +1,339 @@
+/*
+
+   The MIT License (MIT)
+
+   Copyright (c) 2017 Tim Warburton, Noel Chalmers, Jesse Chan, Ali Karakus
+
+   Permission is hereby granted, free of charge, to any person obtaining a copy
+   of this software and associated documentation files (the "Software"), to deal
+   in the Software without restriction, including without limitation the rights
+   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+   copies of the Software, and to permit persons to whom the Software is
+   furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included in all
+   copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+   SOFTWARE.
+
+ */
+
+
+// In the PCG, we collect the alpha/beta from PCG to form tri-diagonal matrix (Lanczo)
+// then the Ritz values will be used to estimated the min/max eigenvalues to later usage
+
+#include "elliptic.h"
+#include "timer.hpp"
+#include "linAlg.hpp"
+
+//#define DEBUG
+static dfloat f77sgn(dfloat val1, dfloat val2) { // return val1 with sign of val2
+  return (val2 < 0) ? -std::abs(val1) : std::abs(val1);
+}
+
+// Solve eigenvalue of tri-diagonal matrix
+// copied from subroutine calc (diag,upper,d,e,n,dmax,dmin) (Nek5000/core/hmholtz.f)
+// TODO: we can call it directly via nekInterface?
+static void calc_eigen(dfloat* diagt, dfloat* upper, const int n, dfloat &dmin, dfloat &dmax)
+{
+  // Rayleight-quotient with QR iter and Wilkinson's shift
+
+  const int verbose = platform->options.compareArgs("VERBOSE", "TRUE");
+  dfloat* d = (dfloat *) calloc(n, sizeof(dfloat));
+  dfloat* e = (dfloat *) calloc(n, sizeof(dfloat));
+
+  memcpy(d, diagt, n*sizeof(dfloat));
+  memcpy(e, upper, n*sizeof(dfloat));
+
+  uint l,m,i,iter,m_in,iskip;
```

Comment:
Prefer declaring variables _as needed_, rather than upfront like this.

For loop variables (e.g., `l`), you can declare them inside the for-loop.

Reply:
In the new version, variable are moved to the placed before it's used.


Additional question here. Which one is better?
```
for (int i=0;i<n;i++)
```
or 
```
int i=0;
for (i=0;i<n;i++)
```

I know the first one is not supported in old standard so I use the second one.